src/variableProcessor.ts

```diff
@@ -1,11 +1,50 @@
 'use strict';
 
-import { window } from 'vscode';
+import { commands, Uri, window } from 'vscode';
 import { EnvironmentController } from './controllers/environmentController';
 import * as Constants from './constants';
 import { Func } from './common/delegates';
 import * as moment from 'moment';
 const uuid = require('node-uuid');
+const adal = require('adal-node');
+const copyPaste = require('copy-paste');
+
+const aadRegexPattern = `\\{\\{\\s*\\${Constants.AzureActiveDirectoryVariableName}(\\s+(${Constants.AzureActiveDirectoryForceNewOption}|${Constants.AzureActiveDirectoryClearCacheOption}))?(\\s+(ppe|public|cn|de|us))?(\\s+([^\\.]+\\.[^\\}\\s]+|[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}))?\\s*\\}\\}`;
+const aadTokenCache = {};
+
+// see UserCodeInfo at https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/blob/dev/lib/adal.d.ts
+interface SignInCode {
+    deviceCode: string;
+    expiresIn: number;
+    interval: number;
+    message: string;
+    userCode: string;
+    verificationUrl: string;
+    error?: any;
+    errorDescription?: any;
+    [x: string]: any;
+}
+
+// see TokenResponse at https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/blob/dev/lib/adal.d.ts
+interface SignInToken {
+    tokenType: string;
+    expiresIn: number;
+    expiresOn: Date | string;
+    resource: string;
+    accessToken: string;
+    refreshToken?: string;

```

Comment:
Would you like to support automatically refreshing access token when expired?

Reply:
Yeah, I put it in my backlog. It'd save a re-auth scenario, so it'd definitely be useful. Not a blocker for basic functionality, tho.